### PR TITLE
[plugins/home] Add Featured Docs Card

### PR DESCRIPTION
**Pull request by nikolarAutodesk**
## Hey, I just made a Pull Request!

Contributing a Home page card we use on our team called Featured Docs. 
This is a simple card that expects a filter to pull whichever entity/s you want to display.
Additional optional props are added to allow for more customization. 

Screenshot 2023-10-03 at 3 47 29 PM

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/home/src/plugin.ts
Comment (awanlin) [inline plugins/home/src/plugin.ts:220]: Can you use the `createCardExtension` instead? Then it can be used in the Customizable Homepage where uses have the ability to add "widgets" and move them around as they like.

File: .changeset/wet-cats-allow.md
Comment (awanlin) [inline .changeset/wet-cats-allow.md:5]: A quick example of the component in use here is super helpful. The changesets get rolled up into release notes so having an example here means Adopters upgrading can start using it right away. Or, you could include a link to the StoryBook entry seeing as you nicely have one :)

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (awanlin) [inline plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:44]: For a future PR: would be good to allow Adopters to customize this

File: plugins/home/src/plugin.ts
Comment (awanlin) [reply plugins/home/src/plugin.ts:220]: Example of this: https://backstage.io/storybook/?path=/story/plugins-home-templates--customizable-template

**Review by awanlin — CHANGES_REQUESTED** [review]
This is a really nice contribution @nikolarAutodesk, just left a few comments 🚀

File: plugins/home/src/plugin.ts
Comment (drodil) [reply plugins/home/src/plugin.ts:220]: If you don't want to render the default card created by `createCardExtension`, you can also use the `createReactExtension` and it works too with the Customizable Homepage.

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (nikolarAutodesk) [reply plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:44]: thank you for the suggestion, I agree this would be good for future!

**Review by awanlin — APPROVED** [review]
Looking good @nikolarAutodesk! 🚀 

Will need a review from the @backstage/discoverability-maintainers now

File: plugins/home/src/plugin.ts
Comment (nikolarAutodesk) [reply plugins/home/src/plugin.ts]: thank you @awanlin & @drodil!
I went with  `createReactExtension` specifically because of types. 
I think there is a few changes that need to be made in `plugins/home-react/src/extensions.tsx` and some other files to avoid ts errors when allowing a `ReactNode` to be passed for the `title` as intended in [InfoCard](https://github.com/backstage/backstage/blob/3cca50a24060efc36164101a80c360c1fe4e3854/packages/core-components/src/layout/InfoCard/InfoCard.tsx#L133). 
We noticed this when we decided to use [StarredEntities](https://github.com/backstage/backstage/blob/3cca50a24060efc36164101a80c360c1fe4e3854/plugins/home/src/plugin.ts#L169) in our own Home Page. We plan on contributing this fix in the future if it is not already under way 👍

File: .changeset/wet-cats-allow.md
Comment (nikolarAutodesk) [reply .changeset/wet-cats-allow.md]: done :)

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (freben) [inline plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:65]: you shouldn't repeat all the types here - no need - typescript extracts them for us. We just do `@param name - desc`.

Even better - if you don't put any params here at all, but instead put those doc strings on individual fields inside `DocsCardProps`, then that type gets fully documented instead which is more correct. Because actually, if you think about it, `FeaturedDocs` has only one single param, namely `props`

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (freben) [inline plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:35]: You need to export and make this one public as well all the way out to the public interface (add it to the index.ts in this folder).

It should also be named after the component it's used in, so `FeaturedDocsProps` I guess?

Remember to run `yarn build:api-reports` and commit the results of that along with your PR.

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (freben) [inline plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:48]: you should ideally always use `theme.spacing` for everything spacing related, for example https://github.com/backstage/backstage/blob/825fcd3f357fd9c9a02d9de19a60d775ff27caa4/packages/core-components/src/components/ErrorPanel/ErrorPanel.tsx#L34

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (freben) [inline plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:92]: It's often best to render the card always, and then have a separate inner `<Content>` component (private) which does all the loading and showing of progress and errors. In a grid of cards, it's confusing to see a free-floating progress bar or warning panel without context of what's going on

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (freben) [inline plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:97]: This should just use ErrorPanel, example: https://github.com/backstage/backstage/blob/825fcd3f357fd9c9a02d9de19a60d775ff27caa4/plugins/catalog-react/src/components/EntityRefLink/FetchedEntityRefLinks.tsx#L91

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (freben) [inline plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:115]: These are all hardcoded routes and we don't know for sure that that's where the docs pages are mounted. Should it instead resolve rootDocsRouteRef? https://github.com/backstage/backstage/blob/825fcd3f357fd9c9a02d9de19a60d775ff27caa4/plugins/techdocs/src/routes.ts#L23

hmmmmm i see it's in a plugin package though instead of in the `-react` package so you should not import that ... ok let's leave this as a future improvement maybe :/

Comment (nikolarAutodesk) [reply plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx]: I added a path parameter as a backup to this.

Comment (nikolarAutodesk) [reply plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx]: Thanks for the review @freben, just pushed the requested changes 👍

**Comment by nikolarAutodesk** [comment]
Bad rebase, working on fixing this

**Comment by nikolarAutodesk** [comment]
@awanlin, I am running into issues resolving these conflicting files. 
Is the appropriate way to update this branch to run: `git rebase origin/master` or `git merge master`?
I also followed instructions [here](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#review-tips) regarding `yarn.lock` conflicts, but pulled in a bunch of unintended changes also. 
Any advice would be helpful 🙏 Thanks in advance!

**Comment by awanlin** [comment]
Hi @nikolarAutodesk, I'm running out of steam for the day, I'll try and look at this tomorrow though 👍

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (tudi2d) [inline plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:95]: Like the intent of customizability. We try to support customisability the following way: https://backstage.io/docs/getting-started/app-custom-theme#overriding-backstage-and-material-ui-components-styles. For this to work follow the documentation adding a `Backstage...` class name.

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (tudi2d) [inline plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:43]: Is there a type defined by Material UI you could use here?

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (tudi2d) [inline plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:119]: Is the default Link color not `primary`? If not, why should it be overwritten?

**Review by tudi2d — CHANGES_REQUESTED** [review]
Thanks for the contribution - appreciate this new Card! Some comments on styling & typing :)

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (nikolarAutodesk) [reply plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx]: I attempted to use a MUI type here, but since I am using other backstage core-components, there is a type error when using another type.

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (nikolarAutodesk) [reply plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:95]: I added more variables in the styles above, but we feel that it increases the flexibility here to allow for `customStyles` to be passed in completely. Can we keep this here?

Comment (tudi2d) [reply plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:95]: Great. OK, let's keep the optional `customStyles`, but please make sure to add name for the component such that people can decide, which option they want to use to override the styles & don't have manually do it in React.

**Review by tudi2d — CHANGES_REQUESTED** [review]
Thanks for adjusting the PR to the review - only a small thing left from my side:) Also: feel free to [follow other companies](https://github.com/backstage/backstage/blob/master/ADOPTERS.md) add [Autodesk in the Adopter List](https://info.backstage.spotify.com/public-adopter-hand-raiser-form) to showcase that you are adopting Backstage. This helps our project:)

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (nikolarAutodesk) [reply plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:98]: @tudi2d I added `{ name: 'BackstageFeaturedDocs' },` 
Please let me know if you meant something else by them 👍

**Comment by nikolarAutodesk** [comment]
Thank you, @tudi2d we are going to add to the Adopters list separately, likely through the form. We need another Autodesk member to approve that on our side 👍

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (vinzscam) [inline plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:105]: small suggestion which consumes less resources backend-side 😉

```suggestion
    const response = await catalogApi.queryEntities({
```

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (vinzscam) [inline plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:150]: according to this logic `<EmptyState />` will be rendered while the entities are loading. It would be nice if we could show `<EmptyState />` once the data has been loaded and there are no matched entities

Comment (nikolarAutodesk) [reply plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:155]: done!

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (nikolarAutodesk) [reply plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:105]: @vinzscam, thank you for the suggestion! I did not know about `queryEntities`, but one thing I did notice from the [API description](https://github.com/backstage/backstage/blob/513bb37fa0033ef7c7a1ac46c04be217cde0bc1d/packages/catalog-client/src/types/api.ts#L522) is that this is set up for paginated responses.
I would have to make some logic updates to handle that case for pagination, but this card is only meant to view  1-5 documents, past that the UI should turn to some sort of [carousel](https://react-bootstrap.netlify.app/docs/components/carousel/) or paginator. 
The current limit for `getEntities` is defaulted to 10 -- is it ok if we leave `getEntities`? 

Either way, I can think of a few places to use this in our implementation at Autodesk 😄 - so thank you!

**Review by emmaindal — APPROVED** [review]
Hello @nikolarAutodesk thanks for your contribution, just reviewed this PR and it looks good to me!

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (vinzscam) [reply plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:105]: I think `queryEntities` should work out of the box, you can ignore all the pagination payload in your case, as you want to retrieve maximum `limit` resources.
But you are right! Since you are passing a fixed limit it is totally fine to use `getEntities` 😄

**Review by vinzscam — APPROVED** [review]
:shipit:

**Comment by awanlin** [comment]
Looks like we have all the approvals? Can someone do the merge? 🎉

**Comment by emmaindal** [comment]
Looks like @backstage-service is blocking us as a codeowner to merge this in 😆 @backstage/maintainers?

File: plugins/home/src/plugin.ts
Comment (freben) [inline plugins/home/src/plugin.ts:221]:
```suggestion
  createComponentExtension({
```

createReactExtension is a low level helper for any react related concerns; actual components use this one

File: plugins/home/src/plugin.ts
Comment (freben) [inline plugins/home/src/plugin.ts:225]: Do you think it might make sense to name this one FeaturedDocsCard (and FeaturedDocsCardProps)? most cards have that suffix. Same for the actual exported variable name on line 220 above of course

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (freben) [inline plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:123]: nitpick: the key should probably be the full entity ref `stringifyEntityRef(d)` since names can be duplicated

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (freben) [inline plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:132]: Hm it's unfortunate to form the link by hand like this, one should ideally use route refs because adopters aren't sure to have their docs page mounted on `/docs` at all.

Unfortunately the docRoot route ref is only exported from the techdocs plugin, not techdocs-react 🤔 https://github.com/backstage/backstage/blob/2cc1dfbaf58ea8a2db2700e60d7a4406ef4ce009/plugins/techdocs/api-report.md?plain=1#L316 that's actually unfortunate because plugins shouldn't import plugins.

OK for now, let's keep this as is, but we should work with @backstage/techdocs-maintainers to move that route ref out to become readable from techdocs-react so we can use it.

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (freben) [inline plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:137]: It's fairly common that descriptions aren't set - should this only be optionally rendered if the description existed?

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.test.tsx
Comment (freben) [inline plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.test.tsx:55]: i noticed that there's an identical testid inside FeaturedDocs, that could probably confuse tests

**Review by freben — APPROVED** [review]
Alright sorry this took some time, a couple of open questions. But at the saame time, the home plugin is on a 0.x version and if you prefer to merge and iterate (even though the props shape might change) later, that's fine too. I'll approve to unblock.

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (nikolarAutodesk) [reply plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:45]: We have found that using the className listed in Backstage, such as `BackstageFeaturedDocs` can lead to uncertainty with design. Meaning, that name can change, and break our designs without us knowing when we update the backstage version (this has happened). 
The thought behind this is to allow users to pass in custom styles without leveraging the override theme functionality. However, this is also based on the class names assigned in FeaturedDocs like `docsTitleLink` - so one could argue it proposes the same risk of breaking changes. It also expects the user to use `makeStyles`
This `customStyles` has gotten a few comments, I can remove it if that is best

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (nikolarAutodesk) [reply plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:43]: Since you can define a primary and secondary color in your global theme, the thought is to allow users to simply specify which color the links should be if not the default `primary`.  
The argument would be that this approach would be a cleaner way to define a color since the hex would only exist in the global theme, and not additionally in the `customStyles` or in the `BackstageFeaturedDocs` override. 

But I can remove this prop since we want people to customize colors using `BackstageFeaturedDocs` if needed

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (nikolarAutodesk) [reply plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:50]: it is the link destination, I can update the name for clarity

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (nikolarAutodesk) [reply plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:49]: this just expands the use case. Currently the path defaults to:
`/docs/${d.metadata.namespace || 'default'}/${d.kind}/${d.metadata.name}/`
so an example is `docs/default/Location/how-to-model-software-in-idp`

But this allows a user to customize that to something else if there is a different entity structure, like:
`/docs/${d.metadata.somethinelse}/${d.kind}/${d.spec.something}/` or `https://backstage.io/docs/getting-started/configuration`

File: plugins/home/src/plugin.ts
Comment (nikolarAutodesk) [reply plugins/home/src/plugin.ts:221]: Ok, I originally had `createComponentExtension` and was suggested to use `createReactExtension` [here](https://github.com/backstage/backstage/pull/20350#discussion_r1347172369)
I'll switch it back

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (nikolarAutodesk) [reply plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:123]: I thought `metatdata.name` must be unique across kinds. Is this documentation outdated? https://backstage.io/docs/features/software-catalog/descriptor-format#name-required

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (nikolarAutodesk) [reply plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx:132]: understood, that is why I added the `path` prop to specify this if it is different.

File: plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx
Comment (freben) [reply plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx]: "per kind, within a given namespace"

So the name + namespace + kind triplet is what needs to be unique. In other words, there may be two users named `freben`, in two different namespaces.

Comment (nikolarAutodesk) [reply plugins/home/src/homePageComponents/FeaturedDocs/FeaturedDocs.tsx]: Got it thank you for clarifying, updated 👍

File: plugins/home/src/plugin.ts
Comment (freben) [reply plugins/home/src/plugin.ts:221]: Hey @nikolarAutodesk I think I'm essentially not entirely sure about the original motivation for the comments that led to this. @drodil said *IF* you want to, etc. But is that really the case? Do you really strongly want a ReactNode title on this or can we just side step that entire problem and keep things as vanilla as possible and use `createCardExtension` which I expect is the originally intended one to leverage? Otherwise it's "just another component" and can't be placed on actual composable home pages.

So I'd like to vote to use `createCardExtension` and sidestep the type issue, then later on work toward ReactNode titles if that's really a feature that's particularly desired. I have my doubts that it's generally a good idea to put things that freely in specifically titles, to be honest.

Comment (Rugvip) [reply plugins/home/src/plugin.ts:221]: What's the reason to not use `createCardExtension`? It's most likely what we want to be able to use here and if there's something preventing that we may want to change `createCardExtension` instead. Otherwise if this is a plain component you should use `createComponentExtension`. Do not use `createReactExtension` as it's a lower level API that we don't want to be used directly, it's for example what's used to implement `createCardExtension` under the hood.

Comment (nikolarAutodesk) [reply plugins/home/src/plugin.ts:221]: @freben @Rugvip thanks for the responses, I can switch to `createCardExtension`, though
our team does strongly leverage the ability to pass in a `ReactNode` title, and if `createCardExtension` uses `InfoCard` ([here](https://github.com/backstage/backstage/blob/3cca50a24060efc36164101a80c360c1fe4e3854/plugins/home-react/src/extensions.tsx#L171)) under the hood, then shouldn't it also allow title to be a `ReactNode` rather than a `string`([here](https://github.com/backstage/backstage/blob/3cca50a24060efc36164101a80c360c1fe4e3854/plugins/home-react/src/extensions.tsx#L82))?

This is my original response for avoiding `createCardExtension`:

> I went with `createReactExtension` specifically because of types.
> I think there are a few changes that need to be made in `plugins/home-react/src/extensions.tsx` and some other files to avoid ts errors when allowing a `ReactNode` to be passed for the title as intended in [InfoCard](https://github.com/backstage/backstage/blob/3cca50a24060efc36164101a80c360c1fe4e3854/packages/core-components/src/layout/InfoCard/InfoCard.tsx#L133).
> We noticed this when we decided to use [StarredEntities](https://github.com/backstage/backstage/blob/3cca50a24060efc36164101a80c360c1fe4e3854/plugins/home/src/plugin.ts#L169) in our own Home Page. We plan on contributing this fix in the future if it is not already under way 👍

Comment (nikolarAutodesk) [reply plugins/home/src/plugin.ts]: updated 👍

**Review by Rugvip — APPROVED** [review]
Looks good, thank you! 👍 

Let's :shipit:, I think we need to treat the title as a separate issue and stick to the card extension.

**Pull request merged**
